Move EventPipe thread buffer allocation outside of lock

josalem opened the pull request:
## Summary

fixes https://github.com/dotnet/runtime/issues/65247

Because GC threads are CPU affinitized, losing a CPU quantum can cause GC delays. Writing an event to EventPipe can incur the need to allocate a new buffer. Currently, that happens behind a lock. When GC related events are turned on, it is possible for GC threads to get blocked waiting on lock long enough to lose their quantum. The extra wait is typically waiting on the allocation and zeroing of the buffer. These buffers are thread local, but need to be added to the global buffer manager behind a lock. To rectify this, this patch moves the allocation of the buffer outside the lock, so there is minimal work inside the lock.

Local testing showed improvement in total throughput (written+dropped) with a higher written count, but not a high enough improvement that I would consider this a performance win beyond the specific GC scenario.

## baseline
```
**** Summary ****
iteration 1: 11,848,091.00 events collected, 28,785,734.00 events dropped in 45.554975 seconds - (29.16% throughput)
	(260,083.36 events/s) (52,016,672.73 bytes/s)


|-----------------------------------|--------------------|--------------------|--------------------|--------------------|
| stat                              | Min                | Max                | Average            | Standard Deviation |
|-----------------------------------|--------------------|--------------------|--------------------|--------------------|
| Events Read                       |       11,848,091.00|       11,848,091.00|       11,848,091.00|                0.00|
| Events Dropped                    |       28,785,734.00|       28,785,734.00|       28,785,734.00|                0.00|
| Throughput Efficiency (%)         |               29.16|               29.16|               29.16|                0.00|
| Event Throughput (events/sec)     |          260,083.36|          260,083.36|          260,083.36|                0.00|
| Data Throughput (Bytes/sec)       |       52,016,672.73|       52,016,672.73|       52,016,672.73|                0.00|
| Duration (seconds)                |           45.554975|           45.554975|           45.554975|            0.000000|
|-----------------------------------|--------------------|--------------------|--------------------|--------------------|
```

## patch
```
**** Summary ****
iteration 1: 12,474,621.00 events collected, 31,134,446.00 events dropped in 45.769763 seconds - (28.61% throughput)
	(272,551.58 events/s) (54,510,315.51 bytes/s)


|-----------------------------------|--------------------|--------------------|--------------------|--------------------|
| stat                              | Min                | Max                | Average            | Standard Deviation |
|-----------------------------------|--------------------|--------------------|--------------------|--------------------|
| Events Read                       |       12,474,621.00|       12,474,621.00|       12,474,621.00|                0.00|
| Events Dropped                    |       31,134,446.00|       31,134,446.00|       31,134,446.00|                0.00|
| Throughput Efficiency (%)         |               28.61|               28.61|               28.61|                0.00|
| Event Throughput (events/sec)     |          272,551.58|          272,551.58|          272,551.58|                0.00|
| Data Throughput (Bytes/sec)       |       54,510,315.51|       54,510,315.51|       54,510,315.51|                0.00|
| Duration (seconds)                |           45.769763|           45.769763|           45.769763|            0.000000|
|-----------------------------------|--------------------|--------------------|--------------------|--------------------|
```

Commit:
Move buffer allocation outside of lock

AaronRobinsonMSFT commented on src/native/eventpipe/ep-buffer-manager.c:
```diff
@@ -501,6 +501,11 @@ buffer_manager_allocate_buffer_for_thread (
+	new_buffer = ep_buffer_alloc (buffer_size, ep_thread_session_state_get_thread (thread_session_state), sequence_number);
+	ep_raise_error_if_nok (new_buffer != NULL);
+
 	// Allocating a buffer requires us to take the lock.
```
This comment seems to indicate an allocation requires the lock. Is this no longer true or is the new allocation not the one the comment is referring to?

josalem replied:
I believe this comment is outdated. Changes in the .net6 timeframe made the reservation of the buffer size lockless and was moved out of the lock. Previously, this action required the lock and so the whole allocation was behind the lock. I'll update the comment.

josalem commented:
The macOS test failure is suspicious:
```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
[createdump] task_for_pid(12343) FAILED 5 (os/kern) failure
/private/tmp/helix/working/B1D809A1/w/A3C508E8/e/tracing/eventpipe/eventsourceerror/eventsourceerror/eventsourceerror.sh: line 396: 12343 Abort trap: 6           (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"

Return code:      1
Raw output file:      /tmp/helix/working/B1D809A1/w/A3C508E8/uploads/Reports/tracing.eventpipe/eventsourceerror/eventsourceerror/eventsourceerror.output.txt
Raw output:
BEGIN EXECUTION
/tmp/helix/working/B1D809A1/p/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false eventsourceerror.dll ''
  0.0s: ==TEST STARTING==
  0.4s: Attempting to delete the zombied pipe: /var/folders/v0/t1tw3f6s2ql4xvj7fjnr0f2h0000gq/T/dotnet-diagnostic-6061-1649806128-socket
  0.4s: Deleted
  0.4s: Attempting to delete the zombied pipe: /var/folders/v0/t1tw3f6s2ql4xvj7fjnr0f2h0000gq/T/dotnet-diagnostic-11189-1649812636-socket
  0.4s: Deleted
  0.4s: Attempting to delete the zombied pipe: /var/folders/v0/t1tw3f6s2ql4xvj7fjnr0f2h0000gq/T/dotnet-diagnostic-9101-1649810377-socket
  0.4s: Deleted
  0.4s: Started sending sentinel events...
  0.4s: Connecting to EventPipe...
  0.6s: Creating EventPipeEventSource...
  0.7s: EventPipeEventSource created
  0.7s: Dynamic.All callback registered
  0.7s: Running optional trace validator
  0.7s: Finished running optional trace validator
  0.7s: Starting stream processing...
  0.9s: Saw sentinel event
  0.9s: Stopped sending sentinel events
  0.9s: Starting event generating action...
  1.0s: Stopping event generating action
  1.0s: Sending StopTracing command...
  1.2s: Saw new provider 'IllegalTypesEventSource'
  1.2s: Saw new provider 'Microsoft-DotNETCore-EventPipe'
  1.2s: Saw new provider 'Microsoft-Windows-DotNETRuntimeRundown'
Expected: 100
Actual: 134
END EXECUTION - FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=/tmp/helix/working/B1D809A1/p
> /private/tmp/helix/working/B1D809A1/w/A3C508E8/e/tracing/eventpipe/eventsourceerror/eventsourceerror/eventsourceerror.sh
```

Based on the evidence, it looks like the main program NullRef'd and _not_ the target process where the code that is changed would be executed. The dump of the process wasn't uploaded because it was "too large" so I'm flying blind right now. I'm running it in a loop locally to try and catch it. I'll rerun the test in CI as well.

> ⚠[DumpFileToLarge: The dump /cores/core.12343 is 6342742016 bytes, which is larger than the supported 1610612736.0, so was not uploaded.]()
> ⚠[DumpFileToLarge: The dump /cores/core.12343 is 6342742016 bytes, which is larger than the supported 1610612736.0, so was not uploaded.]()

CC @dotnet/runtime-infrastructure - Is there any way to get the dumps referenced in the Helix Artifacts tab or are they lost to the void since they weren't uploaded?

CC @mikem8361 - Is there a reason the dumps for these simple tests are ~6GB? That seems excessively large and is preventing the dumps from being uploaded by the test infra.

Commit:
Update comment

josalem commented:
/azp run runtime

hoyosjs commented:
/azp run

hoyosjs commented:
/azp run runtime

tommcdon commented:
> CC @dotnet/runtime-infrastructure - Is there any way to get the dumps referenced in the Helix Artifacts tab or are they lost to the void since they weren't uploaded?

My guess is that they are lost.  The dump file settings can be found here: https://dev.azure.com/dnceng/internal/_git/dotnet-helix-machines?path=/resources/helix-scripts/helix/executor.py&version=GBmain&line=722&lineEnd=727&lineStartColumn=1&lineEndColumn=33&lineStyle=plain&_a=contents.  The infra puts two restrictions for dumps - first is a per-dump restriction, and the other is a total number of dumps threshold.  The dump thresholds depend on the machine type.  For on-prem machines, per-dump is 1.5 GB and the total limit is 2 GB.  Otherwise the per-dump limit is 8 GB and the total is 12 GB.  It feels like the limits are reasonable and so we should find a way to reduce the dump sizes or find a different way of emitting diagnostic data.
 
> CC @mikem8361 - Is there a reason the dumps for these simple tests are ~6GB? That seems excessively large and is preventing the dumps from being uploaded by the test infra.

My guess is that there is a bug in the dump generation code where we are needlessly writing reserved but not actively used pages to disk.  In the short term perhaps we should default to mini dumps for macos.  @mikem8361 do we have a bug tracking this for net7?
Another idea is that we could use createdump's report feature to dump a json representation of the crash.  That might be enough for a reasonable number of investigations to be unblocked.

mikem8361 commented:
>> CC @mikem8361 - Is there a reason the dumps for these simple tests are ~6GB? That seems excessively large and is preventing the dumps from being uploaded by the test infra.

>My guess is that there is a bug in the dump generation code where we are needlessly writing reserved but not actively used >pages to disk. In the short term perhaps we should default to mini dumps for macos. @mikem8361 do we have a bug >tracking this for net7?

If this dump generated by createdump, we have issue #68604 tracking that. If it is generated by the system/MacOS, there is nothing we can do about that,

hoyosjs commented:
The logs are gone - this code can grab dumps with heap https://github.com/dotnet/runtime/blob/1351ac873087a33865370c2eb6b0e29a898dacb0/src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs#L209-L224, but there's also the question if it was a system dump. 

@josalem is this ready to go in?

josalem commented:
> is this ready to go in?

I think so. I want a few code reviews after seeing that one-off AV on the mac queue. I couldn't repro it and it hasn't happened in CI since.

noahfalk reviewed (approved):
LGTM

mikem8361 commented:
>>Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
>>
>> [createdump] task_for_pid(12343) FAILED 5 (os/kern) failure

>> CC @mikem8361 - Is there a reason the dumps for these simple tests are ~6GB? That seems excessively large and is preventing the dumps from being uploaded by the test infra.

It looks like the dumps are still being generated by the system and there seems to be no way that I know to control the size.  It also looks like createdump is also setup to generate the core dump but fails probably because it isn’t signed/entitled (and the host i.e corerun would also need to be) since this is a CI build.  I’m not sure what to do about this createdump problem yet.
